content/guest-post/session-based-authentication-with-nestjs-and-mongodb/index.md

```diff
@@ -0,0 +1,977 @@
+
+So, create a new file in the auth folder for serializer and deserializer, name it `session.serializer.ts`.
+
+At this point, the shape of our application `src` folder should look like this.
+
+```
+    └───src
+    │   └───auth
+    │   │   └───auth.module.ts
+    │   │   └───auth.service.ts
+    │   │   └───authenticated.guard.ts
+    │   │   └───local.auth.guard.ts
+    │   │   └───local.strategy.ts
+    │   │   └───session.serializer.ts
+    │   └───users
+    │   │   └───users.controller.ts
+    │   │   └───users.model.ts
+    │   │   └───users.module.ts
+    │   │   └───users.service.ts
+    │   └───app.controller.ts
+    │   └───app.module.ts
+    │   └───app.service.ts
+    │   └───main.ts
+```
+
+`session.serializer.ts`
+
+```js
+import { Injectable } from "@nestjs/common"
+import { PassportSerializer } from "@nestjs/passport"
+
+@Injectable()
+export class SessionSerializer extends PassportSerializer {
+  serializeUser(user: any, done: (err: Error, user: any) => void): any {
+    done(null, user)
+  }
+  deserializeUser(
+    payload: any,
+    done: (err: Error, payload: string) => void
+  ): any {
+    done(null, payload)
+  }
+}
+```
+
+Go back to your `auth.module.ts` file, provide the `SessionSerializer` and add the `register` method to the `PassportModule`.
+
+`auth.module.ts`
+
+```js
+import { Module } from "@nestjs/common"
+import { PassportModule } from "@nestjs/passport"
+import { UsersModule } from "src/users/users.module"
+import { AuthService } from "./auth.service"
+import { LocalStrategy } from "./local.strategy"
+import { SessionSerializer } from "./session.serializer"
+
+@Module({
+  imports: [UsersModule, PassportModule.register({ session: true })],
+  providers: [AuthService, LocalStrategy, SessionSerializer],
+})
+export class AuthModule {}
+```
+
+Within the `LocalAuthGuard` in the `local.auth.guard.ts` file, add some codes.
+
+Call the `login` method in `super`, pass in the request to trigger the actual login via creating a session. If you want to use sessions, you must remember to trigger the `super.login()`.
+
+`local.auth.guard.ts`
+
+```js
+    import { ExecutionContext, Injectable } from '@nestjs/common';
+    import { AuthGuard } from '@nestjs/passport';
+    @Injectable()
+    export class LocalAuthGuard extends AuthGuard('local') {
+      async canActivate(context: ExecutionContext) {
+        const result = (await super.canActivate(context)) as boolean;
+        const request = context.switchToHttp().getRequest();
+        await super.logIn(request);
+        return result;
+      }
+    }
+```
+
+If you log in now, you would see the session ID stored in a cookie, which is just a key to the session store, and the cookie gets saved in the browser. The cookie is automatically attached to the rest of the request.
+
+Now that the session is working, you can access the protected route, it would return the expected user’s details.
```

Comment:
Confirm if the `AuthService` was included in the `auth.module.ts`.
Also I had to follow my tutorial from begining and everything works.

Reply:
Confirm if the `AuthService` was included in the `auth.module.ts`.
Also I had to follow my tutorial from begining and everything works.

Reply:
@icode247   I have added `AuthModule` in `app.module.ts` file because of that i got that issue. now that issue is fixed. 

Now Signup is working fine at my end. But when i run '/users/login', I get `internal server error` and the application is crashing too.

![image](https://user-images.githubusercontent.com/16396865/163803863-660e4aeb-0100-435b-8a56-a1d8b3b6b0ea.png)
![image](https://user-images.githubusercontent.com/16396865/163803985-d5fb4341-a774-4a3a-9e8e-4e79606e1ff0.png)

Reply:
Please cross check the code you have on your `/auth` folder.
It works on my own end too

Reply:
Please follow the tutrial more carefully or crose check the code you have with the one on repo included in the draft.